io-engine/src/bdev/nexus/nexus_channel.rs

```diff
@@ -193,67 +196,69 @@ impl<'n> NexusChannel<'n> {
-
-        let mut writers = Vec::new();
-        let mut readers = Vec::new();
-
-        // iterate over all our children which are in the healthy state
-        self.nexus()
-            .children_iter()
-            .filter(|c| c.is_healthy())
-            .for_each(|c| match (c.get_io_handle(), c.get_io_handle()) {
-                (Ok(w), Ok(r)) => {
-                    writers.push(w);
-                    readers.push(r);
-                }
-                _ => {
-                    c.set_faulted_state(FaultReason::CantOpen);
-                    error!("{self:?}: failed to get I/O handle for {c:?}");
-                }
-            });
-
-        // then add write-only children
-        if !readers.is_empty() {
-            self.nexus()
-                .children_iter()
-                .filter(|c| c.is_rebuilding())
-                .for_each(|c| match c.get_io_handle() {
-                    Ok(hdl) => {
-                        debug!(
-                            "{self:?}: connecting child device \
-                                in write-only mode: {c:?}"
-                        );
-                        writers.push(hdl);
+        match dev_ctx {
+            DrEvent::ChildRebuildBegin(dev) => {
+                if let Some(child) =
+                    self.nexus().child_by_device_name(dev.as_str())
+                {
+                    if !self.readers.is_empty() {
+                        // TODO: assert that the device is !is_writer here.
+                        match child.get_io_handle() {
+                            Ok(hdl) => {
+                                debug!("{self:?}: connecting child device in write-only mode: {child:?}");
+                                self.writers.push(hdl);
+                            }
+                            Err(e) => {
+                                child.set_faulted_state(FaultReason::CantOpen);
+                                error!("{self:?}: failed to get I/O handle for {child:?}: {e}");
+                            }
+                        }
+                        // Make sure the child is not in readers. Most likely
+                        // the child won't be in readers
+                        // already here.
+                        self.readers
+                            .retain(|h| &h.get_device().device_name() != dev);
                     }
-                    Err(e) => {
-                        c.set_faulted_state(FaultReason::CantOpen);
-                        error!(
-                            "{self:?}: failed to get I/O handle \
-                                for {c:?}: {e}"
-                        );
+                }
+            }
+            DrEvent::ChildRebuildEnd(dev) => {
+                if let Some(child) =
+                    self.nexus().child_by_device_name(dev.as_str())
+                {
+                    if child.is_healthy() {
+                        // TODO: assert that the device is_writer and not
```

Comment:
Should we check that child is in the self.writers, and if not don't add it to the readers?

Reply:
we can check that but I think that should be an assertion (as TODO says), because it shouldn't be possible - that the rebuild finished and child is healthy but it's not a channel writer on this core?

Reply:
It can happen because calling reconfigure is async, so things may change underneath you.